pgmpy/inference/mplp.py

```diff
@@ -0,0 +1,752 @@
+import itertools
+import numpy as np
+from operator import sub
+from operator import add
+import queue
+import math
+from collections import namedtuple
+from pgmpy.models import MarkovModel
+from pgmpy.inference import ApproxInference
+
+
+class MPLP(ApproxInference):
+    """
+    Class for performing inference using Max-Product Linear Programming Relaxations.
+    We use the Max-Product Linear Programming (MPLP) algorithm to minimize the dual .
+    """
+
+    def __init__(self, model):
+        super().__init__(model)
+
+        if not isinstance(model, MarkovModel):
+            raise ValueError('The model given is not MarkovModel.')
+
+        # For containing the cardinalities of all the variables present in the model
+        self.m_var_sizes = self.cardinality
+        # All the factors are kept here
+        self.m_all_intersect = self.factors
+        # All the messages per factor is kept here. It is updated in each iteration of run_mplp
+        self.objective = []
+        # It contain the Region objects for all the factors which are made of more than one variable
+        self.m_all_regions = []
+        # Initial messages for more than single node regions are kept here. These don't change.
+        self.m_region_lambdas = []
+        # Maximum integral solution of the single nodes makes the m_best_val
+        self.m_best_val = -1000000
+        # This is to be used for calculating the breaking point of RunMPLP
+        self.last_obj = 1e40
+        self.m_intersect_map = []
+        for i in range(len(self.factors)):
+            self.objective.append([np.log(y) for y in self.factor_values[i]])
```

Comment:
Why store all the factor values in another variable again? Don't create ```factor_values``` class. You can change the code to 

```python
for factor in self.factors:
    self.objective.append(np.log(factor.values))
```

I think it would serve the same purpose and there is no need to create another list to store all the values. I strongly suggest you to minimize the use of list as much as possible; they are not memory efficient.